### PR TITLE
Issue #7398: Fix groovy + java-library plugin compatibility

### PR DESCRIPTION
**Pull request by JamesXNelson**
### Context

Per issue #7398, using the java-library plugin with the groovy plugin causes groovy output to not be transitively included (when referenced as a project dependency).

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#git-commit---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

**Comment by big-guy** [comment]
@ljacomet are you going to get to this for 5.3 still?

**Comment by ljacomet** [comment]
This looks good indeed and should be integrated indeed. Of course, we left it for the last minute.
@JamesXNelson Any chance you can fix that DCO issue and rebase it all on top of the `release` branch then edit this PR to get it targeted at `release`?

**Comment by ljacomet** [comment]
I have cleared this from Gradle 5.3. The tests are failing on Windows, there is some file locking issue that prevent test state to be deleted.
Sorry it took so long to get attention. However the changes are OK so there is only work to be done around the tests and fixing the sign off from the documentation commit.

**Comment by ljacomet** [comment]
@JamesXNelson 5.3 RC1 has been postponed a bit. If you can get to this before the weekend, it has a chance of going in

**Comment by JamesXNelson** [comment]
I'll take a look now and try to get everything resolved tonight.  Going to have to wait until nightly OSS session after kids are asleep; can't spend company time on it.

**Comment by JamesXNelson** [comment]
@ljacomet ready to go.  I update the docs to mention 5.3 as the line where it's fixed "for free".  

Likely wanna delete the section mentioning the workaround _eventually_; will also want to verify what happens when someone with the publishing workaround inherits the fix from upstream, in case it needs to be called out in release notes.

**Comment by JamesXNelson** [comment]
is there anything todo for me w.r.t. the failed checks above?  I do not have visibility into the details.

**Comment by blindpirate** [comment]
@JamesXNelson you can see the failure details by "logging as guest".

**Comment by ljacomet** [comment]
Sadly the added tests still have issues on Windows - see https://builds.gradle.org/viewLog.html?buildId=20670097&buildTypeId=Gradle_Check_Stage_ReadyforMerge_Trigger

Looks like something done in the tests prevents clean up on Windows.

**Comment by JamesXNelson** [comment]
login as guest does indeed work.  I'll try to get this in tonight, once the kids are asleep.

Actually, I think I can guess... windows machines prefer jar over classes when using java-library, right?

If so, then I think the test just needs to update it's assertions to account for OS.  What's the best way to get the OS in the test fixture?  I'm assuming it is correct to assert that "windows always jar", "unix will use classes dirs when it can" ?

**Comment by melix** [comment]
> Actually, I think I can guess... windows machines prefer jar over classes when using java-library, right?

No, they don't. This is something we're discussing, to mitigate a performance problem under windows, but that's not the case today.

**Comment by JamesXNelson** [comment]
Aye.  I've realized now that I've looked at the test...  the compile classpath is using java-api-jars, so it's never actually referencing directories on any OS.

I'll update / simplify the test to reflect this.  Would you rather I simply delete the excessive test cases that no longer show anything of interest, or leave the test there to break if the current-default behavior changes?

My own $0.02 is that it would be nice to choose prefer-classes over prefer-jars, but that's mostly because I maintain compiler plugins / annotation processors which already know how to infer source locations from classes directories...  Always using jars seems to erase the benefit of "only consider java classes" for speedy multi-module compiles, but I didn't measure anything, so just let me know how much of this integration test I should delete to get this merged. :-)

**Comment by melix** [comment]
>  I've realized now that I've looked at the test... the compile classpath is using java-api-jars, so it's never actually referencing directories on any OS.

The consumer requests `java-api` (via `compileClasspath`). A disambiguation rule will automatically prefer `classes` over `java-api-jars`, so it shouldn't be a problem.

The test which fails under windows is failing because it tries to delete the temporary directory after the build completed. And I think the problem is that in your test you create a `JarFile` and get an entry, but never close this file, so the test fails.

**Comment by JamesXNelson** [comment]
Ah, thanks.  I'll try it again.  

I just ran it on my linux box, and it failed here:

```
assert compileJava.classpath.files.size() == 2
assert compileJava.classpath.files.containsAll(classesDirs)
```
as the classpath on the compileJava task had only the jar in the `files`.

**Comment by JamesXNelson** [comment]
(I already pushed amended commit which is probably not correct yet)

**Comment by JamesXNelson** [comment]
ack.  that build is gonna die.  so sorry... ...it's ...quite late here.

**Comment by melix** [comment]
I can take over the PR, since your commit is signed and has DCO. Are you find with this?

**Comment by JamesXNelson** [comment]
sure.  I don't mind figuring it out, since I'm doing other work that will be affected by these changes, but it sounds like you will be able to fix it faster than me.

In the test, the resolved variant winds up w/ usage of `java-api-jars`, and `org.gradle.dependency.bundling=external`, when java is depending on groovy.  
It passed when the original commit was authored :-)

**Comment by JamesXNelson** [comment]
image

File: subprojects/docs/src/docs/userguide/java_library_plugin.adoc [hunk 1]
```diff
@@ -237,15 +237,14 @@ The role of each configuration is described in the following tables:
...
+At the moment, the Java Library plugin is wired to behave correctly with the `java`, `groovy` and `kotlin` plugins. In Gradle 5.3 or early, some plugins, such as the Groovy plugin, may not behave correctly. In particular, if the Groovy plugin is used in addition to the `java-library` plugin, then consumers may not get the Groovy classes when they consume the library. To workaround this, you need to explicitly wire the Groovy compile dependency, like this:
 
-.Configuring the Groovy plugin to work with Java Library
+.Configuring the Groovy plugin to work with Java Library in Gradle 5.3 or earlier
 ====
 include::sample[dir="java-library/with-groovy/groovy",files="a/build.gradle[tags=configure-groovy]"]
 include::sample[dir="java-library/with-groovy/kotlin",files="a/build.gradle.kts[tags=configure-groovy]"]
 ====
 
-
```
Comment (JamesXNelson) [inline subprojects/docs/src/docs/userguide/java_library_plugin.adoc]: woops.  I deleted I line I added encouraging users to combine groovy + java-library to get the "use classes dir" behavior that is no longer relevant / true.

If this newline is significant, it should be restored.

**Comment by JamesXNelson** [comment]
I pushed the missing `where:` header and jar file bits at least.  will check back tomorrow after work! ty.

File: subprojects/docs/src/docs/userguide/java_library_plugin.adoc [hunk 2]
```diff
@@ -237,15 +237,14 @@ The role of each configuration is described in the following tables:
 [[sec:java_library_known_issues_compat]]
 === Compatibility with other plugins
 
-At the moment the Java Library plugin is only wired to behave correctly with the `java` plugin. Other plugins, such as the Groovy plugin, may not behave correctly. In particular, if the Groovy plugin is used in addition to the `java-library` plugin, then consumers may not get the Groovy classes when they consume the library. To workaround this, you need to explicitly wire the Groovy compile dependency, like this:
+At the moment, the Java Library plugin is wired to behave correctly with the `java`, `groovy` and `kotlin` plugins. In Gradle 5.3 or early, some plugins, such as the Groovy plugin, may not behave correctly. In particular, if the Groovy plugin is used in addition to the `java-library` plugin, then consumers may not get the Groovy classes when they consume the library. To workaround this, you need to explicitly wire the Groovy compile dependency, like this:
```
Comment (JLLeitschuh) [inline subprojects/docs/src/docs/userguide/java_library_plugin.adoc]:
```suggestion
At the moment, the Java Library plugin is wired to behave correctly with the `java`, `groovy` and `kotlin` plugins. In Gradle 5.3 or earlier, some plugins, such as the Groovy plugin, may not behave correctly. In particular, if the Groovy plugin is used in addition to the `java-library` plugin, then consumers may not get the Groovy classes when they consume the library. To workaround this, you need to explicitly wire the Groovy compile dependency, like this:
```

**Comment by JLLeitschuh** [comment]
Should Kotlin also be fixed so this works or should that fix be included in the Kotlin plugin source code instead of Gradle patching for it?

**Comment by melix** [comment]
AFAIK the Kotlin plugin is already compatible

**Comment by JamesXNelson** [comment]
Perhaps at a later time, the integration test instead becomes "JavaPluginCompatibilityIntegrationTest", and all jvm platforms can be checked for expected behavior.  Reading your new and improved version has been very enlightening; ty.

File: subprojects/plugins/src/integTest/groovy/org/gradle/groovy/GroovyJavaLibraryInteractionIntegrationTest.groovy [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+            }
+            file('javaLib').with {
+                file('src/main/java/JavaClass.java') << "public class JavaClass { GroovyClass reference; }"
+                file('build.gradle') << """
+                        apply plugin: '$consumerPlugin'
+                        dependencies {
+                          $consumerConf project(':groovyLib')
+                        }
+                        
+                        task assertDependsOnGroovyLib {
```
Comment (melix) [inline subprojects/plugins/src/integTest/groovy/org/gradle/groovy/GroovyJavaLibraryInteractionIntegrationTest.groovy]: Actually this whole block can be removed, the test doesn't use it anymore

**Comment by JLLeitschuh** [comment]
@JamesXNelson Thanks for signing off that PR for me. I approve that sign off on my behalf.

**Comment by ljacomet** [comment]
@melix This now passes a full build. Is it good to merge? We have the opportunity of pushing it to RC2 if that's the case.

**Comment by melix** [comment]
There's a minor unresolved comment, otherwise lgtm

**Comment by ljacomet** [comment]
I took care of that comment, will merge if build remains OK.

**Comment by JamesXNelson** [comment]
@JLLeitschuh sorry, I am not used to adding sign-offs to pull requests.  I assumed that since the request came from my fork that I had to sign any commits coming from not-gradle-team-members (i.e. unsure if you had permission to force-push a signoff as well).  For future pull requests, I'll only signoff my own work. :-)

**Comment by JamesXNelson** [comment]
And thanks for getting this over the finish line, I'm in the middle of an international move, so not really able to find much extra time to pay proper attention.

**Comment by JLLeitschuh** [comment]
@JamesXNelson It's totally fine. Thank you for doing it. I don't mind at all.

**Pull request merged**